Using PR labels to handle Review Apps (#3)
fastruby/RFCs: 2021-09-30-review-apps-github-labels.md

# Handling Review Apps using labels in GitHub PRs

## Problem:

> Heroku's Review Apps are created automatically in many projects when a PR is created/updated. While this helps in the QA process, this is has downsides in some cases:

Reviewer: ```suggestion
Heroku's Review Apps are created automatically in many projects when a PR is created/updated. While this helps in the QA process, this has downsides in some cases:
```

- if the PR is a work-in-progress, the review app is probably not needed yet
- if the PR is waiting for a review of the code, the review app is probably not needed yet

> - if the PR changes something that can't really be QA'd wit ha review app, it's not needed at all (like updating the readme, setting up docker, fixing a broken test, etc)

Reviewer: ```suggestion
- if the PR changes something that can't really be QA'd with a review app, it's not needed at all (like updating the readme, setting up docker, fixing a broken test, etc)
```

> - Automatically-created Review Apps are deleted after 2 days so the person doing QA has to go to heroku and re-create the review app manually if it's done after that time

Reviewer: ```suggestion
- Automatically-created Review Apps are deleted after 2 days so the person doing QA has to go to heroku and re-create the review app manually if it's more than 2 days since the Review App has been created
```

In all those cases, we have Review Apps running for a couple of days that nobody is using, costing money, and when somebody actually needs to use them, the review app is already destroyed and the re-creating process involves many steps.

## Proposal:

> Use Github Workflows to trigger Review Apps creation/destructions from within the PR by using specific labels. This way, the Review Apps won't be created automatically if not needed (saving money), and when somebody is ready to QA they don't need to go to Heroku and remember all the steps to create them manually (saving time).

Reviewer: 👍

- Add labels to a PR https://github.com/TimonVS/pr-labeler-action (this includes connecting to the GitHub API to set labels for a PR, we can use it as reference for removing the label instead)
- How to create Github Actions https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action (this shows the step-by-step on creating github actions)
- Documentation on setting ENV variables for the GitHub Actions https://docs.github.com/en/actions/learn-github-actions/environment-variables (needed for the Heroku's API token and Pipeline ID)

## Complete workflow:

- A developer creates a PR (no review app is created)
- When the PR is ready for QA and the person doing the QA is ready, they will add the `deploy-review-app` label to the PR
- After a moment, the GitHub Action will be triggered, creating a Review App and removing the `deploy-review-app` label (the information of the deploy should be added to the PR by Heroku like it's currently doing)

> - When the QA is done, the QA person will add the `destroy-review-app` label to the PR

Reviewer 2: I like how this sounds and I think it makes sense to start this way. But I also see a lot of room for human error here. We've had problems before with people forgetting the Ready to Merge label, so there's reason to believe they'd forget this one as well.

Might be worth exploring for a future iteration here if we can integrate Jira into this mix. And then once the story goes on to `Done` the review app can be destroyed regardless of label?

Author: the review app will be destroyed when the PR is merged or closed with an automation, regarding of the `destroy-review-app` label, the label is there to destroy the review app manually if you want to

also thinking about open source projects using this, that's why we didn't put jira in the mix

- After a moment, the GitHub Action will be triggered, destroying the Review App and removing the `destroy-review-app` label

### Extra steps:

> - We could have a workflow configuration that destroys the review app when a `ready-to-merge` label is added (so the QA person don't need to remember to add the `destroy-review-app` label in that case).

Reviewer 3: ```suggestion
- We could have a workflow configuration that destroys the review app when a `ready-to-merge` label is added (so the QA person doesn't need to remember to add the `destroy-review-app` label in that case).
```

Minor grammar fix

> - We could have a workflow configuration that destroys the review app when a `ready-to-merge` label is added (so the QA person doesn't need to remember to add the `destroy-review-app` label in that case).

Reviewer 2: @arielj I really like the overall idea, but to be honest I'd like to move away from labels like the Ready to Merge one, it should be a column in the Jira workflow instead.

Author: though not all project will have a jira project (like points)

> - We should have a workflow configuration that destroys the review app when the PR is closes (either close or merge+close) to prevent stale review apps running when not needed anymore.

Reviewer 3: ```suggestion
- We should have a workflow configuration that destroys the review app when the PR is closed (either close or merge+close) to prevent stale review apps running when not needed anymore.
```

> In these 2 cases, we would re-use the previous GitHub Action that destroy the Review App by just setting other jobs for the specific triggers. Here's an example of a GitHub Workflows job when a PR is closed https://github.com/arielj/test-gh-labels/blob/main/.github/workflows/on_closed.yml.

Reviewer 3: ```suggestion
In these 2 cases, we would re-use the previous GitHub Action that destroys the Review App by just setting other jobs for the specific triggers. Here's an example of a GitHub Workflows job when a PR is closed https://github.com/arielj/test-gh-labels/blob/main/.github/workflows/on_closed.yml.
```

# Final thoughts:

> The code of the GitHub Actions will be public and the different projects will reference them, so it's reusable. The only code that would be copied on each project is the workflows `.yml` file that sets the triggers.

Reviewer: Does this mean there would be a repo for these actions and the implementing project just needs a reference to the repo with the GithubActions "code"?

Reviewer 4: That would be cool because then we could open source the action itself!

Author: > Does this mean there would be a repo for these actions and the implementing project just needs a reference to the repo with the GithubActions "code"?

yes, that's how actions are created, you create a public repo following the specs and just add `uses: fastruby/create-review-app-from-label@v1` on the steps of the job and github will get that repo and run the code

like this https://github.com/mheap/github-action-pr-heroku-review-app
